MSC2783: Homeserver Migration Data Format (#2783)
matrix-org/matrix-spec-proposals: proposals/2783-homeserver-migration-data-format.md

become compliant, there isn't much holding back to migrate to these more performant implementations, with the
aforementioned "lossy" migration method being the only way, which could result in many "dead domains" and a hard breakage
of rooms and DMs for many.

## Proposal

This proposal aims to provide a better, lock-in free, and reliable way of migrating between implementations,
as well as be able to provide a format in which a "snapshot" backup of a server could be made.

This is supplemental to the existing spec, and this does not aim to influence the general protocol spec,
but rather provide a standard "mold" to "pour" a homeserver's data into.

This proposal aims to be extendable by custom homeserver implementations and other interested parties or documents,
be it that some data is implementation-specific, or that data is custom to MSCs or other "extended specifications".

(This proposal uses [RFC2119](https://tools.ietf.org/html/rfc2119) to indicate requirement levels. e.g. "SHOULD", "MUST", etc.)

## General Structure

The proposal mainly defines a directory structure, this directory structure can be captures in ZIP files,

> RAR files, `.tar.gz` files, or any other sort of archival or indexable directory "target".

Author: All of the above formats can be interpreted as a file system directory structure, and I'm not going to mandate which one to pick, but to allow any sort of "input" to the importer

The `mspf` here stands for "Matrix Server Persistance Format", this is not a custom file format,
but is suffixed before `.json` to not confuse this file with other "`manifest.json`"-identified formats.

The manifest contains information pertaining to "items", items can concern themselves with things such as room events,
queued `to_device` events, user account data, but also custom data (such as data pertaining to SSO)

All items are mapped with their java-domain-notation specifier (e.g. `org.example.subdomain`), all items are JSON `object`s.

The Item `object`s can be freely fitted for their purpose, but every Item SHOULD have their version specifier specified in the key `v`,
which can be any JSON value itself (though `int`, `array[int]` or `string` are recommended.)

All items starting their specifier with `m.` MUST be documented by the matrix spec itself, and any other specification SHALL NOT
claim a specifier with this prefix (such as MSCs and custom implementations).

However, also, when processing a manifest, *all* items prefixing with `m.` MUST be processed or otherwise handled,

> when an importer encounters a `m.`-prefixed item specifier it does not understand, it must abort the import process.

Reviewer: This seems strong. What if an old deprecated feature is not supported. I think ideally the operator would be notified and given the choice.

I think the spec isn't the right place to decide that some imports MUST be aborted. In general for transfers like this "as much as possible with a report of failures" is what I want to see.

Author: "As much as possible" is lossy, I explicitly don't want a lossy import, the `m.`-prefixed keys mandate *core data structures*, ones that *cannot* be ignored, such as events, keys, login details, and other such things.

Any other key is best-effort, in *that* namespace you can add impl-specific config or data, and there it's free game, but not with core data.

Author: Core data is core data, if you lose out on core data, no matter what it is, your homeserver will still have scars, so no.

Lossiness will only apply to the non-`m.`-prefixed areas, because non-`m.` is "best effort only", but missing out a `m.` is unacceptable (you wouldn't want user accounts to be malformatted when imported, or entirely missing, would you?).

> However, also, when processing a manifest, *all* items prefixing with `m.` MUST be processed or otherwise handled,
> when an importer encounters a `m.`-prefixed item specifier it does not understand, it must abort the import process.

Reviewer 2: Do not couple the importer and service semantics:
```suggestion
However, also, when processing a manifest, *all* items prefixing with `m.` MUST be processed or otherwise handled,
when an importer encounters a `m.`-prefixed item specifier it does not understand, it must go on regardless.
```

Author: No. Core data is core data.

Items that come with version 0 of the manifest (and version 1 when standardized in the spec).

### `m.core` version `1`

`m.core` aims to capture the absolute "core" items of a matrix server.

It owns the file `m.core.json`, with the following structure;

XXX: Get more feedback if this is all "core" data.
```yaml
{
  # The servername part of a matrix server, required.
  "server_name": "example.com"

  # The signing key of the server.
  "signing_key": "ed25519 [...]"
}
```

### `m.rooms` version `1`

`m.rooms` aims to capture room metadata such as versions, memberships, and aliases.

It owns a the file `m.rooms.cbor`.

The file contains a mapping (`{}`) of room ID -> room details,
room details is another mapping with the following structure;

XXX: Aliases need better looking at.
```yaml
{
  # Room version, a string
  version: "1",

  # Local aliases to this room, array of strings
  aliases: [
    "#room:example.com"
  ]
}
```

### `m.events` version `1`

`m.events` aims to capture all events present in the database.

This includes local rooms, unfederated rooms, encrypted rooms, etc.

It owns the directory `m.events/`, and files `events.*.cbor`.

> The files contain CBOR-encoded mappings of room ID -> array of events.

Reviewer: The spec is all JSON except for https://github.com/matrix-org/matrix-doc/pull/3079 which isn't accepted yet. Should we just stick with JSON for now to ensure that there are no ambiguities?

Author: JSON is way too sparse of a data structure, costly without reason, CBOR is a direct alternative without compromises.

Though, I'm probably going to change this to SQLite databases because CBOR files of this type don't have iterative parsers/consumers for every language out there, and I want memory usage to be low, and not that a 1 GB CBOR file causes a 2 GB memory spike on top of it.

Reviewer 3: Directly mapped JSON to CBOR has minimal size advantages. For a 352 byte standard event:
```
{"type":"m.room.member","sender":"@username:matrix.org","content":{"avatar_url":"mxc://matrix.org/FLOIFFxSxJVrtqHqKJAvaZGNRJ","membership":"join","displayname":"Username"},"event_id":"$15943751151sWhiO:matrix.org","unsigned":{"age":14,"replaces_state":"$15943751120CQQql:matrix.org"},"state_key":"@username:matrix.org","origin_server_ts":1594375115332}
```

- As direct CBOR: 308 bytes
- As gzipped JSON: 259 bytes
- As MSC3079 CBOR: 205 bytes
- As gzipped MSC3079 CBOR: 178 bytes

Author: just for curiosity's sake, what about gzipped CBOR?

Reviewer 3: 254 bytes.

Event formats are room-version specific, and so for this proposal, they're opaque `object`s.

Event ordering isn't guaranteed, and it is not even guaranteed that all events to a room are only
saved on one file.

XXX: RoomID -> Array[Event] mapping because just making it arrays could be a lot of overhead for importers,
this way the importers can just "select" rooms from files and correctly parse version-specific event formats
from them, revert? or keep?

### `m.users` version `1`

`m.users` intends to capture all user-specific information.

This includes account data, room tags, and user devices.

It owns the directory `m.users/`, and files `users.*.cbor`.

The files contain CBOR-encoded mappings of user ID -> user details.

> A user ID key mapping MUST only exist *once* across all files.

Reviewer: Why? A server can have a lot of user info and it may be useful to split it into multiple files to enable parallel dumping/loading. For example imagine a bot user that is in many thousands of rooms. This also seems inconsistent with the `m.rooms` export.

Author: User details contain information about account data and keys and such, user membership is in the state events given with the regular "event data", and a server can/should reconstruct membership based on that data, like the old server can.

  password_hash: "",

  # A UNIX timestamp for when the user has created this account.
  created_at: 0,

  # A 'roomID (string) -> {tag (string) -> content (object)}' mapping
  room_tags: {
    "!abc:example.org": {
      "m.tag": {}
    }
  },

  # A mapping of DeviceID -> Device
  devices: {
    "DEADBEEF": {
      # A freeform string, can be null
      display_name: "UwU Matrix Tesseract client (Tesla)",

      # A boolean for if the device is hidden from user profiles or not.
      # XXX: I picked this from the synapse postgres DB table, feedback?
      hidden: false
    }
  }
}
```

### `m.e2e.to_device` version `1`

`m.e2e.to_device` aims to capture all in-flight/undelivered `to_device` events.

XXX: I'm not sure if this should be under `e2e`, thoughts?

XXX: Does this also have to include outbound `to_device` events that havent been delivered yet?

It owns the files `to_device.*.cbor` under directory `m.e2e/`.

XXX: TODO; structure of files, array? mapping?

### `m.e2e.keys` version `1`

`m.e2e.keys` aims to capture all E2E-encryption-key related data.

XXX: Need expertise for this, I don't know how much or what specifically i should or could capture here.

> ## Potential issues

Reviewer: Import object size.

1. Large files may be hard for servers to manage. Ideally we would allow non-streaming parsing to be viable for ease of client implementations. Maybe we need to limit files to be <16MiB or something? (Note that the in-memory representation of COBR/JSON can be multiple times larger than the serialized representation).
1. Large files that must be parsed sequentially may prevent parallelism and cause unacceptably slow imports. For example large files are are a single JSON or COBR object can not be parsed in parallel. Maybe we should consider a format such as JSON objects on separate lines which can be parsed in parallel. (That being said tar archives can't fully be parsed in parallel anyways, but they can be streamed and the import can be parallelized)

Author: 1. Yes, large object files are bad, but many small files are also bad, as that is relatively heavy on the file system, so I decided to experiment with SQLite for the time being.
2. The parallelisation needs to be at both ends, and I don't know if a server importer could support importing multiple events from multiple timelines at once, maybe "per room", but even then for the stateres (for it to be efficient) it has to happen sequentially per room.

Reviewer 2: > The parallelisation needs to be at both ends, and I don't know if a server importer could support importing multiple events from multiple timelines at once, maybe "per room", but even then for the stateres (for it to be efficient) it has to happen sequentially per room.

If you treat the import as a bulk data input, then you can do the state resolution after completing the import. Do not couple the import and processing.
